packages/hardhat-core/src/internal/util/download.ts

```diff
@@ -27,47 +21,48 @@ export async function download(
   timeoutMillis = 10000
 ) {
   const { pipeline } = await import("stream");
-  const { default: fetch } = await import("node-fetch");
+  const { getGlobalDispatcher, ProxyAgent, request } = await import("undici");
   const streamPipeline = util.promisify(pipeline);
-  const fetchOptions: FetchOptions = {
-    timeout: timeoutMillis,
-    agent: undefined,
-  };
 
-  // Check if Proxy is set https
-  if (process.env.HTTPS_PROXY !== undefined) {
-    // Create the proxy from the environment variables
-    const proxy: string = process.env.HTTPS_PROXY;
-    fetchOptions.agent = new HttpsProxyAgent.HttpsProxyAgent(proxy);
-  }
+  const dispatcher = (() => {
+    if (process.env.HTTPS_PROXY !== undefined) {
+      return new ProxyAgent(process.env.HTTPS_PROXY);
+    }
 
-  // Check if Proxy is set http and `fetchOptions.agent` was not already set for https
-  if (
-    process.env.HTTP_PROXY !== undefined &&
-    fetchOptions.agent === undefined
-  ) {
-    // Create the proxy from the environment variables
-    const proxy: string = process.env.HTTP_PROXY;
-    fetchOptions.agent = new HttpsProxyAgent.HttpsProxyAgent(proxy);
-  }
+    if (process.env.HTTP_PROXY !== undefined) {
+      return new ProxyAgent(process.env.HTTP_PROXY);
+    }
+
+    return getGlobalDispatcher();
+  })();
 
   // Fetch the url
-  const response = await fetch(url, fetchOptions);
+  const response = await request(url, {
+    dispatcher,
+    headersTimeout: timeoutMillis,
+    maxRedirections: 10,
+    method: "GET",
+    headers: {
+      "User-Agent": "hardhat",
```

Comment:
Nice!

What if we add the version number here? It can be useful for debugging.

Reply:
Let's ignore this for now. I thought this was the HTTP provider, which would be much more useful.

We can add that in a future PR

Reply:
Had already drafted it, so pushed it up. Addressed in https://github.com/NomicFoundation/hardhat/pull/2370/commits/fd875048a9437dea109e654b44331679d7d3e45e